File: python/Q66_plus_one.py
#
# @lc app=leetcode id=66 lang=python3
#
# [66] Plus One
#


# @lc code=start
class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        carry: int = 1
        result: list[int] = []

        for d in reversed(digits):
            if carry:
                if d + carry == 10:
                    result.append(0)
                else:
                    result.append(d + carry)
                    carry = 0
            else:
                result.append(d)

        if carry:
            result.append(carry)

        result.reverse()

        return result


# @lc code=end

File: rust/src/leetcode/Q66_plus_one.rs
/*
 * @lc app=leetcode id=66 lang=rust
 *
 * [66] Plus One
 */

// @lc code=start
impl Solution {
    pub fn plus_one(digits: Vec<i32>) -> Vec<i32> {
        let mut result: Vec<i32> = digits;

        for i in (0..result.len()).rev() {
            if result[i] < 9 {
                result[i] += 1;
                return result;
            } else {
                result[i] = 0;
            }
        }

        let mut extended_result: Vec<i32> = vec![1];
        extended_result.extend(result);
        extended_result
    }
}
// @lc code=end
